contrib/grpcplugins/deployment/plugin-kubernetes/main.go

```diff
@@ -0,0 +1,469 @@
+	return &platformplugin.PlatformPluginManifest{
+		Name:        "Kubernetes Deployment Plugin",
+		Author:      "Benjamin Coenen",
+		Description: "Kubernetes Deployment Plugin",
+		Version:     sdk.VERSION,
+	}, nil
+}
+
+// getStartingConfig implements ConfigAccess
+func getStartingConfig(token, timeout string) clientcmd.KubeconfigGetter {
+	return func() (*clientcmdapi.Config, error) {
+		defaultClientConfigRules := clientcmd.NewDefaultClientConfigLoadingRules()
+		overrideCfg := clientcmd.ConfigOverrides{
+			AuthInfo: clientcmdapi.AuthInfo{
+				Token: token,
+			},
+			Timeout: timeout + "s",
+		}
+
+		clientConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(defaultClientConfigRules, &overrideCfg)
+		rawConfig, err := clientConfig.RawConfig()
+		if os.IsNotExist(err) {
+			return clientcmdapi.NewConfig(), nil
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		return &rawConfig, nil
+	}
+}
+
+func (k8sPlugin *kubernetesDeploymentPlugin) Deploy(ctx context.Context, q *platformplugin.DeployQuery) (*platformplugin.DeployResult, error) {
+	k8sAPIURL := q.GetOptions()["cds.platform.api_url"]
+	k8sToken := q.GetOptions()["cds.platform.token"]
+	k8sCaCertificate := q.GetOptions()["cds.platform.ca_certificate"]
+	deploymentFilepath := q.GetOptions()["cds.platform.deployment_files"]
+	helmChart := q.GetOptions()["cds.platform.helm_chart"]
+
+	if k8sToken == "" {
+		return fail("Kubernetes token should not be empty")
+	}
+
+	certb64 := base64.StdEncoding.EncodeToString([]byte(k8sCaCertificate))
+	kubecfg := fmt.Sprintf(`apiVersion: v1
+kind: Config
+users:
+- name: cds
+  user:
+    token: %s
+clusters:
+- cluster:
+    certificate-authority-data: %s
+    server: %s
+  name: self-hosted-cluster
+contexts:
+- context:
+    cluster: self-hosted-cluster
+    user: cds
+  name: default-context
+current-context: default-context`, k8sToken, certb64, k8sAPIURL)
+
+	if err := os.Mkdir(".kube", 0755); err != nil {
+		return fail("Cannot create directory .kube : %v", err)
+	}
+	defer func() {
+		if err := os.RemoveAll(".kube"); err != nil {
+			fmt.Printf("Cannot delete .kube directory : %v\n", err)
+		}
+	}()
+
+	if err := ioutil.WriteFile(".kube/config", []byte(kubecfg), 0755); err != nil {
+		return fail("Cannot write kubeconfig : %v", err)
+	}
+
+	switch {
+	case helmChart != "":
+		if err := executeHelm(q); err != nil {
+			return fail(err.Error())
+		}
+	case deploymentFilepath != "":
+		if err := executeK8s(q); err != nil {
+			return fail(err.Error())
+		}
+	default:
+		return fail("Must have deployment_files or helm_chart not empty")
+	}
+
+	return &platformplugin.DeployResult{
+		Status: sdk.StatusSuccess.String(),
+	}, nil
+}
+
+func (k8sPlugin *kubernetesDeploymentPlugin) DeployStatus(ctx context.Context, q *platformplugin.DeployStatusQuery) (*platformplugin.DeployResult, error) {
+	return &platformplugin.DeployResult{
```

Comment:
Do we have an API to know the exact status of the deployment?

Reply:
I used the flag --wait to let kubectl wait until all deployments are done. So as the arsenal plugin it's not required